### PR TITLE
Tpsa discretization class

### PR DESCRIPTION
**Pull request by keileg**
## Proposed changes
This PR introduces a two-point stress discretization for the elasticity problem. Tests are also provided.

An integration of the discretization method into the multiphysics models is in the works, but is not included in this PR. Also, a tutorial documenting usage may be included at some point, but for now, the documentation is in the form of a MVE inside the tpsa class.

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.
- [x] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: tests/numerics/fv/test_tpsa.py [hunk 1]
```diff
@@ -0,0 +1,1458 @@
...
+    Returns:
+        np.ndarray: Array of boundary condition values, as a g.dim x num_faces array.
+
+    """
+
+    face_ind = g.get_all_boundary_faces()
+    nf = face_ind.size
+
+    # Find the faces on the boundary in each direction
+    min_coord, max_coord = pp.domain.grid_minmax_coordinates(g)
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:1409]: Please consider using function in test_mpsa (moving it to applications?)

File: src/porepy/numerics/fv/tpsa.py [hunk 2]
```diff
@@ -0,0 +1,1379 @@
...
+
+        # Boundary conditions for the rotation variable. This should only be used if
+        # the Cosserat parameter is non-zero. Since the rotation variable is scalar if
+        # nd == 2 and vector if nd == 3, the type of boundary condition depends on the
+        # dimension.
+        bnd_rot: pp.BoundaryCondition | pp.BoundaryConditionVectorial = (
+            parameter_dictionary.get("bc_rot", None)
+        )
+
+        # Check that the type of boundary condition is consistent with the dimension.
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:551]: That would be convenient,  if it integrates reasonably well with the existing code. User project?

File: src/porepy/numerics/fv/tpsa.py [hunk 3]
```diff
@@ -0,0 +1,1379 @@
...
+        # correct scaling with the face area. Explicitly filter Neumann boundary
+        # conditions, these do not contribute.
+        stress_total_pressure = (
+            filters.neu_nopass_nd @ normal_vector_diag @ c2f_maps.c2f_compl_scalar_2_nd
+        )
+
+        # The solid mass conservation equation is discretized by taking the average
+        # displacement over the faces and scaling with the normal vector. To that end,
+        # construct a sparse matrix that has one normal vector per row.
+        normal_vector_nd = sps.csr_matrix(
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:687]: Agree, it is at least similar. The Ad extension of tpsa should be very similar to AdTpfa, at least in spirit, although I don't see us prioritizing this in the near future.

File: src/porepy/numerics/fv/tpsa.py [hunk 4]
```diff
@@ -0,0 +1,1379 @@
...
+    mu_by_dist_fc_cc: np.ndarray
+    """Shear modulus divided by the distance from face to cell centers. Follows the
+    fi/ci ordering defined in _Numbering."""
+    mu_by_dist_fc_cc_bound: np.ndarray
+    """Shear modulus divided by the distance from face to cell centers, also accounting
+    for Robin boundary conditions. Follows the fi/ci ordering defined in _Numbering."""
+    inv_mu_by_dist_array: sps.dia_array
+    """Diagonal array of the inverse of mu_by_dist_fc_cc_bound."""
+    rob_weight: np.ndarray
+    """Robin weights, extracted from the displacement boundary conditions."""
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:121]: The Robin condition can be interpreted as an elastic modulus divided by a distance measure, and this is the terminology used in the Tpsa paper (indeed, stare at the relevant parts of the paper for some days, and it is amazing what you can get used to). Thus, in the context of that description of the discretization, it is natural to consider the Robin weight a distance, and include it here.

I will include an explanatory remark, though; the question is a fair one.

File: src/porepy/numerics/fv/tpsa.py [hunk 5]
```diff
@@ -0,0 +1,1379 @@
...
+        self.stress_displacement_matrix_key: str = "stress"
+        """Keyword used to identify the discretization matrix for the stress generated
+        by the cell center displacements. Defaults to 'stress'.
+        """
+        self.stress_rotation_matrix_key: str = "stress_rotation"
+        """Keyword used to identify the discretization matrix for the rotational stress
+        generated by the cell center rotational stress. Defaults to
+        'stress_rotation'."""
+        self.stress_total_pressure_matrix_key: str = "stress_total_pressure"
+        """Keyword used to identify the discretization matrix for the solid mass
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:265]: Cut and paste error. Thanks.

File: src/porepy/numerics/fv/tpsa.py [hunk 6]
```diff
@@ -0,0 +1,1379 @@
...
+        #       b(\sigma_k * n_k - g^u) = 2\mu (g^u - u_k).
+        #
+        # Here, b is considered a length scale and is also denoted \delta_k^j (see the
+        # bottom of page 10 in the arXiv version of the paper, link above). It is
+        # understood that b = 0 gives a Dirichlet condition, b = infinity gives a
+        # Neumann condition, and other values (positive and negative) form the Robin
+        # case. In PorePy, Dirichlet and Neumann conditions are set separately, while
+        # the Robin parameter, \alpha, is interpreted to be on the form
+        #
+        #       \sigma_k * n_k + \alpha u = g^u
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:465]: This is the notation used in the paper, I'll add a comment.

File: src/porepy/numerics/fv/tpsa.py [hunk 7]
```diff
@@ -0,0 +1,1379 @@
...
+        #    larger degree, the more explicit treatment herein was preferred partly for
+        #    ease of implementation and interpretation, and (honestly) partly for
+        #    reasons of exhaustion on EK's part. For future reference, the most natural
+        #    alternative approach is to introduce one weight each for Dirichlet and
+        #    Neumann boundaries and let Robin be an average. The first step would be to
+        #    verify that this does not introduce division by zero somewhere.
+        #
+        # Known shortcomings and known unknowns:
+        # 1. The implementation of boundary conditions for the rotation variable covers
+        #    only Neumann and Dirichlet conditions. Robin conditions should be feasible,
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:488]: For future reference: it will most likely be rather straightforward now that a full implementation of boundary conditions is available, provided the right design choices are made for the implementation. It is not fully clear to me what that caveat  means in practice.

File: src/porepy/numerics/fv/tpsa.py [hunk 8]
```diff
@@ -0,0 +1,1379 @@
...
+        )
+
+        #################
+        # Start construction of discretization matrices.
+
+        # Compute t_shear as the sum of the contribution from the interior faces
+        # (computed here) and the Robin boundary conditions (represented in t_shear_rob,
+        # computed previously). Since the Robin conditions are already handled, we use
+        # mu_by_dist_fc_cc, not mu_by_dist_fc_cc_bound.
+        t_shear_nd = (
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:651]: The only magic happens in the bincount, which cannot reasonably be split. I see the point about apparent complexity, though, so I added inline comments.

File: src/porepy/numerics/fv/tpsa.py [hunk 9]
```diff
@@ -0,0 +1,1379 @@
...
+            filters.neu_rob_pass_nd @ inv_mu_face
+            - filters.dir_pass_nd
+            - c2f_maps.b2f_rob
+        )
+
+        # Boundary condition for the mass conservation equation. There should be no
+        # contribution from Dofs which are assigned a Dirichlet condition, so filter out
+        # these variables (this is likely not fully consistent for domains with
+        # boundaries not aligned with the coordinate axes, and with rolling boundary
+        # conditions, but EK does not know what to do there).
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:875]: Comment was outdated. Deleted.

File: src/porepy/numerics/fv/tpsa.py [hunk 10]
```diff
@@ -0,0 +1,1379 @@
...
+        sgn_area_scaling = sps.dia_array(
+            (np.repeat(numbering.sgn_bf / sd.face_areas, nd), 0),
+            shape=(nd * nf, nd * nf),
+        )
+
+        # The cell displacement is used for both Robin and Neumann conditions. The
+        # reconstruction can be seen as having two parts: The cell center value
+        # (bound_displacement_cell) and the difference between face and cell, generated
+        # by the different sources of stress and converted to a displacement through an
+        # inversion of Hook's law (the other variables).
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:910]: Specified.

File: src/porepy/numerics/fv/tpsa.py [hunk 11]
```diff
@@ -0,0 +1,1379 @@
...
+        )
+        # For the complement, we only need the diagnoal data.
+        b2f_rob_compl = 1 - b2f_rob.diagonal()
+
+        # Map from scalar cell quantities to nd face quantities (used e.g. for the solid
+        # pressure).
+        c2f_scalar_2_nd = dist.inv_mu_by_dist_array @ sps.kron(
+            cell_to_face, sps.csr_array(np.ones((nd, 1))), format="csr"
+        )
+        # Specifically set a zero value for faces that have Dirichlet conditions, as
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:1128]: I think the chances of them making such a change are slim, and I am confident it will result in failure of tests. Other implementations must be possible, but I don't see an approach that will not be much more cumbersome. So in total, though I agree you have a good point, I am comfortable with keeping this approach.

File: src/porepy/numerics/fv/tpsa.py [hunk 12]
```diff
@@ -0,0 +1,1379 @@
...
+        # (used in the paper to describe the discretization scheme).
+        mu_by_dist_fc_cc_bound = np.bincount(
+            np.hstack((numbering.fi_expanded, rob_boundary_faces_expanded)),
+            weights=np.hstack((2 * mu_by_dist_fc_cc_nd, rob_weights_boundary_faces)),
+        )
+
+        # Create a diagonal matrix that can be used to scale face-to-cell maps to have
+        # unit row sum (thus they become true averaging maps), both in the interior and
+        # on faces with Robin boundary conditions.
+        inv_mu_by_dist_array = sps.dia_array(
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:1238]: I don't think that would work, since we need to left-multiply various matrices with this variable. I am not in complete control of the nuances in scipy-numpy compatability, and associated broadcasting, so my preference is to make it safe.

File: src/porepy/numerics/fv/tpsa.py [hunk 13]
```diff
@@ -0,0 +1,1379 @@
...
+        nc, nf, nd = sd.num_cells, sd.num_faces, sd.dim
+
+        # Get the types of boundary conditions.
+        dir_faces = bnd.is_dir
+        neu_faces = bnd.is_neu
+        rob_faces = bnd.is_rob
+
+        # Data structure for the discretization of the boundary conditions
+        trm_bnd = np.zeros((nd, nf))
+        # On Dirichlet faces, the coefficient of the boundary condition is the same as
```
[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:1299]: Rephrased, will hopefully be better.

File: tests/numerics/fv/test_tpsa.py [hunk 14]
```diff
@@ -0,0 +1,1458 @@
...
+        case _:
+            raise ValueError(f"Unknown boundary condition type {bc_type}")
+
+    d[pp.PARAMETERS][KEYWORD]["bc"] = bc_disp
+    if include_rot:
+        d[pp.PARAMETERS][KEYWORD]["bc_rot"] = bc_rot
+
+
+class TestTpsaTailoredGrid:
+    """Define a cell of two grids, one internal face. Verify that the discretization
```
[the remark this answers is not in the document]
Comment (keileg) [reply tests/numerics/fv/test_tpsa.py:82]: Haha.

File: tests/numerics/fv/test_tpsa.py [hunk 15]
```diff
@@ -0,0 +1,1458 @@
...
+        cosserat = np.array([self.cos_0, self.cos_1])
+
+        self.data = {
+            pp.PARAMETERS: {
+                KEYWORD: {"fourth_order_tensor": C, "cosserat_parameter": cosserat}
+            },
+            pp.DISCRETIZATION_MATRICES: {KEYWORD: {}},
+        }
+
+        # We test the discretization on face 6, as this has two non-trivial components
```
[the remark this answers is not in the document]
Comment (keileg) [reply tests/numerics/fv/test_tpsa.py:133]: Better?

File: src/porepy/numerics/fv/tpsa.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (IvarStefansson) [reply src/porepy/numerics/fv/tpsa.py:687]: I agree. One could think of ploughing through all FV discretisations and extracting common functionality to fv_utils. But again, not a priority, I suspect.

[the remark this answers is not in the document]
Comment (keileg) [reply src/porepy/numerics/fv/tpsa.py:687]: Indeed.

File: src/porepy/numerics/fv/tpsa.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (IvarStefansson) [reply src/porepy/numerics/fv/tpsa.py:551]: Yes. Steps:
1. create issue
2. core dev assesses feasibility
3. initiate user project

File: tests/numerics/fv/test_tpsa.py [hunk 16]
```diff
@@ -550,7 +551,7 @@ def test_robin_bcs(self):
 
         matrices = _discretize_get_matrices(self.g, self.data)
 
-        # The Robin condition translated to 'distances' as is used in the paper.
+        #
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:554]: Delete?

File: tests/numerics/fv/test_tpsa.py [hunk 17]
```diff
@@ -1007,23 +1007,25 @@ def test_cosserat_3d():
...
     inconsistent, and anything can happen), in which case the face normal vectors will
-    point into the domain on the south boundary, out of the domain on the east boundary.
+    point into the domain on the south (and bottom) boundary, out of the domain on the
+    east boundary.
 
     Parameters:
-        g: Grid object.
-        driving_bc_type: Type of boundary condition to apply.
-        extension: If True, the boundary conditions are reversed, such that the force is
+        g: Grid object. driving_bc_type: Type of boundary condition to apply. tensile:
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:1027]: Please move "tensile" to next line.

File: src/porepy/numerics/fv/tpsa.py [hunk 18]
```diff
@@ -0,0 +1,1379 @@
...
+    c2f: sps.sparray
+    """Map from cell to face quantities. Boundary conditions are ignored."""
+    c2f_compl: sps.sparray
+    """Complement map from cell to faces."""
+    c2f_scalar_2_nd: sps.sparray
+    """Map from scalar cell-wise quantities to vector face quantities."""
+    c2f_compl_scalar_2_nd: sps.sparray
+    """Complement map from scalar cell-wise quantities to vector face quantities."""
+    b2f_rob: sps.sparray
+    """Map from faces with Robin boundary conditions asigned to boundary faces. This
```
[the remark this answers is not in the document]
Comment (IvarStefansson) [reply src/porepy/numerics/fv/tpsa.py]: Also, please replace asign* with with assign*

File: tests/numerics/fv/test_tpsa.py [hunk 19]
```diff
@@ -0,0 +1,1454 @@
...
+
+
+@pytest.mark.parametrize("g", [pp.CartGrid([2, 2]), pp.CartGrid([2, 2, 2])])
+@pytest.mark.parametrize("driving_bc_type", ["dir", "neu"])
+@pytest.mark.parametrize("tensile", [False, True])
+def test_compression_tension(g: pp.Grid, driving_bc_type: str, tensile: bool):
+    """Assign a compressive or tensile force, check that the resulting displacement is
+    in the correct direction, and that the total pressure is negative.
+
+    The logic of the tets is as follows (assuming a compressive regime, tensile=False,
```
Comment (IvarStefansson) [inline tests/numerics/fv/test_tpsa.py:1015]:
```suggestion
    The logic of the test is as follows (assuming a compressive regime, tensile=False,
```

File: tests/numerics/fv/test_tpsa.py [hunk 20]
```diff
@@ -0,0 +1,1458 @@
...
+    negative x-direction and positive y- (and z) direction. The total pressure should be
+    negative, while EK cannot surmise the correct sign of the rotation by physical
+    intuition. The grid is assumed to be Cartesian (or else the discretization is
+    inconsistent, and anything can happen), in which case the face normal vectors will
+    point into the domain on the south boundary, out of the domain on the east boundary.
+
+    Parameters:
+        g: Grid object.
+        driving_bc_type: Type of boundary condition to apply.
+        extension: If True, the boundary conditions are reversed, such that the force is
```
[the remark this answers is not in the document]
Comment (IvarStefansson) [reply tests/numerics/fv/test_tpsa.py]: Yes!

**Review by IvarStefansson — COMMENTED** [review]
Almost there!

File: tests/numerics/fv/test_tpsa.py (hunk 1 above)
Comment (keileg) [reply tests/numerics/fv/test_tpsa.py:1403]: Good point. Will make a function in pp.geometry.domains.

**Review by IvarStefansson — APPROVED** [review]
Feel free to merge once all conversations are resolved.
